### PR TITLE
Fix Makefile dependencies for compilerlibs archives

### PR DESCRIPTION
**Pull request by gasche**
compilerlibs archives: add dependencies on .mli-only compilation units

This fixes the issue reported in https://github.com/ocaml/ocaml/pull/2268#issuecomment-568916560.

Before this commit, the Makefile for compilerlibs archives only tracks
their dependencies on implementation object files (.cmo,
.cmx). .mli-only files in those directories result in a .cmi without
a .cmo.

The main makefile uses dependencies on compilerlibs archives as
a proxy for all dependencies on the corresponding compiler
modules. This assumption was broken by .mli-only modules not being
dependencies on the archives.

For example, rebuilding compilerlibs archives would not refresh
file_format/cmxs_format.cmi. The following steps would lead to a build
failure:

    make core
    make ocamltools
    <add a new blank line at the beginning of utils/misc.mli>
    make ocamltools

    File "/home/gasche/Prog/ocaml/github-trunk/tools/objinfo.ml", line 1:
    Error: The files /home/gasche/Prog/ocaml/github-trunk/utils/misc.cmi
           and /home/gasche/Prog/ocaml/github-trunk/file_formats/cmxs_format.cmi
           make inconsistent assumptions over interface Misc

(The first commit in this PR moves the Makefile targets for compilerlibs archives to a separate, included Makefile.)

### CODE REVIEW

File: compilerlibs/Makefile.compilerlibs [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+  asmcomp/debug/available_regs.cmo \
+  asmcomp/debug/compute_ranges_intf.cmo \
+  asmcomp/debug/compute_ranges.cmo \
+  asmcomp/schedgen.cmo asmcomp/scheduling.cmo \
+  asmcomp/branch_relaxation_intf.cmo \
+  asmcomp/branch_relaxation.cmo \
+  asmcomp/emitaux.cmo asmcomp/emit.cmo asmcomp/asmgen.cmo \
+  asmcomp/asmlink.cmo asmcomp/asmlibrarian.cmo asmcomp/asmpackager.cmo \
+  driver/opterrors.cmo driver/optcompile.cmo
+ASMCOMP_CMI=$(INTEL_SPECIFIC_ASMCOMP_CMI)
```
Comment (lthls) [inline compilerlibs/Makefile.compilerlibs]: Isn't this a typo for `ARCH_SPECIFIC_ASMCOMP_CMI` ?

File: compilerlibs/Makefile.compilerlibs [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+  middle_end/flambda/flambda_invariants.cmo \
+  middle_end/flambda/traverse_for_exported_symbols.cmo \
+  middle_end/flambda/build_export_info.cmo \
+  middle_end/flambda/closure_offsets.cmo \
+  middle_end/flambda/un_anf.cmo \
+  middle_end/flambda/flambda_to_clambda.cmo \
+  middle_end/flambda/flambda_middle_end.cmo
+MIDDLE_END_FLAMBDA_CMI=\
+  middle_end/flambda/inlining_decision_intf.mli \
+  middle_end/flambda/simplify_boxed_integer_ops_intf.mli
```
Comment (lthls) [inline compilerlibs/Makefile.compilerlibs]: I assume you meant `.cmi`

File: compilerlibs/Makefile.compilerlibs [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+  middle_end/flambda/effect_analysis.cmo \
+  middle_end/flambda/inlining_cost.cmo \
+  middle_end/flambda/simple_value_approx.cmo \
+  middle_end/flambda/export_info.cmo \
+  middle_end/flambda/export_info_for_pack.cmo \
+  middle_end/compilenv.cmo \
+  $(MIDDLE_END_CLOSURE) \
+  $(MIDDLE_END_FLAMBDA)
+MIDDLE_END_CMI=\
+  middle_end/backend_intf.mli \
```
Comment (lthls) [inline compilerlibs/Makefile.compilerlibs]: `.cmi`

File: compilerlibs/Makefile.compilerlibs [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+  driver/makedepend.cmo \
+  driver/compile_common.cmo
+COMP_CMI=\
+  file_formats/cmo_format.cmi \
+  file_formats/cmx_format.cmi \
+  file_formats/cmxs_format.cmi
+# cmxs_format is required for native compilation (which would suggest
+# a placement in ASMCOMP), but it is also a dependency of ocamlobjinfo
+# that we want to be able to build on bytecode-only systems if
+# possible.
```
Comment (lthls) [inline compilerlibs/Makefile.compilerlibs]: Technically, this is true also of `cmx_format`. Given that we want `ocamlobjinfo` to know about as many ocaml files as possible, I would rephrase it to "All file format descriptions are in the ocamlcommon library so that ocamlobjinfo can depend on them" or something similar. Also, I was wondering why `cmi_format` and `cmt_format` were missing, but since they have implementations they're actually included earlier.

File: compilerlibs/Makefile.compilerlibs [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+$(COMMON:.cmo=.cmx) $(BYTECOMP:.cmo=.cmx) $(OPTCOMP:.cmo=.cmx): ocamlopt
+$(OPTTOPLEVEL:.cmo=.cmx): ocamlopt
+
+
+compilerlibs/ocamlcommon.cma: $(COMMON_CMI) $(COMMON)
+	$(CAMLC) -a -linkall -o $@ $(COMMON)
+partialclean::
+	rm -f compilerlibs/ocamlcommon.cma
+
+compilerlibs/ocamlcommon.cmxa: $(COMMON:.cmo=.cmx)
```
Comment (lthls) [inline compilerlibs/Makefile.compilerlibs]: You probably forgot `COMMON_CMI`

**Review by lthls — COMMENTED** [review]
I have found a few likely typos.
I like this approach to explicitly list mli-only files somewhere, and splitting in a different Makefile seems also a good idea to me.

I wish this file could be generated or checked at least semi-automatically (there's no way to notice if you've put the cmo files in the wrong order until you get a not very helpful link error), but that's clearly beyond the scope of this PR.

I haven't tried to check that no `make` targets are broken by this PR, but if CI passes I'm fine with this getting merged.

**Comment by gasche** [comment]
Thanks a lot! (Apologies for the typos.)

I agree that an automatic analysis would be nice. We want to ensure that all compiler compilation-units end up listed in one of those archive-creating variables, and that the split between mli-only and normal units is correct.

For the record, I used the following command to list all `.mli` files that may belong to mli-only modules (only an approximation due to `.mly`, directory-include tricks etc.):

```
for f in $(find . -name '*.mli' | grep -v testsuite); do
    if test ! -e $(echo $f | sed s/\\.mli/\\.ml/); then
        echo $f;
    fi;
done
```

File: compilerlibs/Makefile.compilerlibs (hunk 1 above)
Comment (gasche) [reply compilerlibs/Makefile.compilerlibs]: It was, thanks!

File: compilerlibs/Makefile.compilerlibs (hunk 4 above)
Comment (gasche) [reply compilerlibs/Makefile.compilerlibs]: I rephrased following the suggestion, thanks.

File: compilerlibs/Makefile.compilerlibs [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+OPTCOMP=$(MIDDLE_END) $(ASMCOMP)
+
+TOPLEVEL=toplevel/genprintval.cmo toplevel/toploop.cmo \
+  toplevel/trace.cmo toplevel/topdirs.cmo toplevel/topmain.cmo
+
+OPTTOPLEVEL=toplevel/genprintval.cmo toplevel/opttoploop.cmo \
+  toplevel/opttopdirs.cmo toplevel/opttopmain.cmo
+
+$(COMMON:.cmo=.cmx) $(BYTECOMP:.cmo=.cmx) $(OPTCOMP:.cmo=.cmx): ocamlopt
+$(OPTTOPLEVEL:.cmo=.cmx): ocamlopt
```
Comment (gasche) [inline compilerlibs/Makefile.compilerlibs:275]: Note: the `OPTTOPLEVEL` line here was missing, I added it for consistency. (It may be a bugfix.)

File: compilerlibs/Makefile.compilerlibs [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+
+compilerlibs/ocamltoplevel.cma: $(TOPLEVEL)
+	$(CAMLC) -a -o $@ $^
+partialclean::
+	rm -f compilerlibs/ocamltoplevel.cma
+
+compilerlibs/ocamlopttoplevel.cmxa: $(OPTTOPLEVEL:.cmo=.cmx)
+	$(CAMLOPT) -a -o $@ $^
+partialclean::
+	rm -f compilerlibs/ocamlopttoplevel.cmxa compilerlibs/ocamlopttoplevel.$(A)
```
Comment (gasche) [inline compilerlibs/Makefile.compilerlibs]: Note: the `$(A)` removal was missing from the previous `partialclean` rule -- this is again a bugfix.

**Comment by gasche** [comment]
@lthls I fixed the typos you found and updated the branch.

File: compilerlibs/Makefile.compilerlibs [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+$(OPTTOPLEVEL:.cmo=.cmx): ocamlopt
+
+
+compilerlibs/ocamlcommon.cma: $(COMMON_CMI) $(COMMON)
+	$(CAMLC) -a -linkall -o $@ $(COMMON)
+partialclean::
+	rm -f compilerlibs/ocamlcommon.cma
+
+compilerlibs/ocamlcommon.cmxa: $(COMMON_CMI) $(COMMON:.cmo=.cmx)
+	$(CAMLOPT) -a -linkall -o $@ $^
```
Comment (lthls) [inline compilerlibs/Makefile.compilerlibs]: Wrong command (causes failures in CI): `$^` should be replaced by `$(COMMON:.cmo=.cmx)`

Comment (gasche) [reply compilerlibs/Makefile.compilerlibs]: Nice catch, thanks. I initially developed an earlier version of this PR (which I believe was correct and typo-free) in the root Makefile, then decided to do the makefile split first, and the rebase introduced a lot of small mistakes which I had fixed earlier... Sorry for the trouble, and thanks for the extra pair of eyes.

Comment (lthls) [reply compilerlibs/Makefile.compilerlibs]: The fix you've pushed is wrong: you're feeding `.cmo`s to `ocamlopt`

**Comment by gasche** [comment]
( @lthls this now seems to have converged to a test-passing state )

**Review by lthls — APPROVED** [review]

**Comment by gasche** [comment]
Thanks!
